src/main/webapp/app/iris/exercise-chatbot/exercise-chatwidget/exercise-chat-widget.component.ts

```diff
@@ -107,6 +109,17 @@ export class ExerciseChatWidgetComponent implements OnInit, OnDestroy, AfterView
         this.dialog.closeAll();
     }
 
+    rateMessage(message_id: number, index: number, helpful: boolean) {
+        this.httpMessageService
+            .rateMessage(<number>this.sessionId, message_id, helpful)
+            .toPromise()
+            .then(() => this.stateStore.dispatch(new RateMessageSuccessAction(index, helpful)))
```

Comment:
I'm more familiar with the idea that states in the store should be immutable. That means every state's changes should be made by dispatching actions. It's also the rule of some famous state store like redux. Do you think we should follow this practice?

Reply:
Thank you for the suggestion! I feel like either way is good, and since it doesn't bring any real benefit to change and I had already implemented by dispatching action, I'm gonna keep it as is :)